amr-wind/boundary_conditions/wall_models/WallFunction.cpp

```diff
@@ -116,14 +128,31 @@ void VelWallFunc::wall_model(
                         const amrex::Real vv =
                             vold_arr(i, j, k + idx_offset, 1);
                         const amrex::Real wspd = std::sqrt(uu * uu + vv * vv);
-                        // Dirichlet BC
-                        varr(i, j, k - 1, 2) = 0.0;
+                        const amrex::Real xc = problo[0] + (i + 0.5) * dx[0];
+                        const int k_lo =
+                            static_cast<int>(amrex::Math::floor(dx[0] / dx[2]));
+                        const int k_hi = k_lo + 1;
+                        AMREX_ALWAYS_ASSERT(
+                            bx.contains(amrex::IntVect(i, j, k_lo)));
+                        AMREX_ALWAYS_ASSERT(
+                            bx.contains(amrex::IntVect(i, j, k_hi)));
+                        const amrex::Real z_diff = dx[0] - (k_lo + 0.5) * dx[2];
```

Comment:
There is still the negative z_diff issue that I mentioned previously. This logic for interpolation needs to be more nuanced to handle several edge cases.

Reply:
The issue stems from the fact that you're using `floor` a ratio of `dx[0]/dx[2]`  and subsequently adding a 0.5 to this floor, I think the actual k_lo calculation should be `floor(dx[0]/dx[2] - 0.5)`.

Please double check to ensure that this inequality is always true: `(k_lo + 0.5) * dx[2] < dx[0] < (k_lo + 1 + 0.5)*dx[2]`, otherwise your interpolation will not work. 

You're still operating under the assumption that dx[0] > dx[2] too, might want to check for that as well.

Reply:
I agree. I just did some modifications to the interpolation to try to make it more general. I think it's a little bit convoluted but i understand why we need to do this.